provider/provider.go

```diff
@@ -213,25 +232,75 @@ func New(opts ...Option) (*SweepingProvider, error) {
 		connectivity.WithOfflineCheckInterval(cfg.connectivityCheckOfflineInterval),
 	)
 	if err != nil {
+		cleanup(prov.cleanupFuncs)
 		return nil, err
 	}
+	prov.cleanupFuncs = append(prov.cleanupFuncs, prov.connectivity.Close)
 
 	go prov.run()
 
 	return prov, nil
 }
 
+const retryInterval = 5 * time.Minute
+
 func (s *SweepingProvider) run() {
+	s.wg.Add(1)
+	defer s.wg.Done()
+
 	logger.Debug("Starting SweepingProvider")
 	go s.measureInitialPrefixLen()
 
-	// TODO: complete me
-	s.handleReprovide() // FIXME: to satisfy linter
+	retryTicker := time.NewTicker(retryInterval)
+	defer retryTicker.Stop()
+
+	for {
+		select {
+		case <-s.done:
+			return
+		case <-retryTicker.C:
+			if s.connectivity.IsOnline() {
+				s.catchupPendingWork()
+			}
+		case <-s.scheduleTimer.C:
+			s.handleReprovide()
+		}
+	}
 }
 
 // Close stops the provider and releases all resources.
-func (s *SweepingProvider) Close() {
-	s.closeOnce.Do(func() { close(s.done) })
+func (s *SweepingProvider) Close() error {
+	var err error
+	s.closeOnce.Do(func() {
+		close(s.done)
+		s.cancelCtx()
+		s.wg.Wait()
+
+		s.scheduleTimer.Stop()
+		err = cleanup(s.cleanupFuncs)
+	})
+	return err
+}
+
+func cleanup(funcs []func() error) error {
+	slices.Reverse(funcs) // LIFO: last-added is cleaned up first
+	var errs []error
+	for _, f := range funcs {
+		if f == nil {
+			continue
+		}
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					errs = append(errs, fmt.Errorf("cleanup panic: %v", r))
+				}
+			}()
+			if err := f(); err != nil {
+				errs = append(errs, err)
+			}
+		}()
```

Comment:
The outer function is not doing anything. Did you intend to call it as a goroutine?
```suggestion
		defer func() {
			if r := recover(); r != nil {
				errs = append(errs, fmt.Errorf("cleanup panic: %v", r))
			}
		}()
		if err := f(); err != nil {
			errs = append(errs, err)
		}
```